polygeist/lib/Dialect/Polygeist/Analysis/SYCLBufferAnalysis.cpp

```diff
@@ -0,0 +1,296 @@
+                           ? subBufOffset
+                           : SmallVector<size_t, 3>{};
+
+    return BufferInformation(jointSize, SubBufferLattice::YES, jointBaseBuf,
+                             jointBaseBufSize, jointOffset);
+  }
+
+  // When control flow reaches this point, not both incoming information are
+  // definitely a sub-buffer. If they are both definitely *not* a sub-buffer,
+  // preserve that fact, otherwise be conservative and assume the resulting
+  // information might be a sub-buffer.
+  auto jointSubBuf =
+      (subBuf == other.subBuf) ? SubBufferLattice::NO : SubBufferLattice::MAYBE;
+
+  return BufferInformation(jointSize, jointSubBuf, nullptr, {}, {});
+}
+
+//===----------------------------------------------------------------------===//
+// SYCLBufferAnalysis
+//===----------------------------------------------------------------------===//
+
+SYCLBufferAnalysis::SYCLBufferAnalysis(Operation *op, AnalysisManager &mgr)
+    : operation(op), am(mgr), idRangeAnalysis(op, mgr) {}
+
+SYCLBufferAnalysis &SYCLBufferAnalysis::initialize(bool useRelaxedAliasing) {
+
+  // Initialize the dataflow solver
+  aliasAnalysis = &am.getAnalysis<mlir::AliasAnalysis>();
+  aliasAnalysis->addAnalysisImplementation(
+      sycl::AliasAnalysis(useRelaxedAliasing));
+
+  // Populate the solver and run the analyses needed by this analysis.
+  solver.load<dataflow::DeadCodeAnalysis>();
+  solver.load<dataflow::SparseConstantPropagation>();
+  solver.load<ReachingDefinitionAnalysis>(*aliasAnalysis);
+
+  if (failed(solver.initializeAndRun(operation))) {
+    operation->emitError("Failed to run required dataflow analyses");
+    return *this;
+  }
+
+  idRangeAnalysis.initialize(useRelaxedAliasing);
+
+  initialized = true;
+
+  return *this;
+}
+
+std::optional<BufferInformation>
+SYCLBufferAnalysis::getBufferInformationFromConstruction(Operation *op,
+                                                         Value operand) {
+  assert(initialized &&
+         "Analysis only available after successful initialization");
+  assert(isa<LLVM::LLVMPointerType>(operand.getType()) &&
+         "Expecting an LLVM pointer");
+
+  const polygeist::ReachingDefinition *reachingDef =
+      solver.lookupState<polygeist::ReachingDefinition>(op);
+  assert(reachingDef && "expected a reaching definition");
+
+  auto mods = reachingDef->getModifiers(operand);
+  if (!mods || mods->empty())
+    return std::nullopt;
+
+  if (!llvm::all_of(*mods,
+                    [&](const Definition &def) { return isConstructor(def); }))
+    return std::nullopt;
+
+  auto pMods = reachingDef->getPotentialModifiers(operand);
+  if (pMods) {
+    if (!llvm::all_of(
+            *pMods, [&](const Definition &def) { return isConstructor(def); }))
+      return std::nullopt;
+  }
+
+  bool first = true;
+  BufferInformation info;
+  for (const Definition &def : *mods) {
+    if (first) {
+      info = getInformation(def);
+      first = false;
+    } else {
```

Comment:
```suggestion
  BufferInformation info = getInformation(mods->front());
  for (const Definition &def : mods->drop_front()) {
```
is something like this doable?

Reply:
`mods` (and `pMods`) is a `std::set` (defined by the `ReachingDefinitionAnalysis`), so there is no `drop_front` and we would need to use a `for`-loop with explicit iterators to achieve this.